src/gc/proxy.d

```diff
@@ -28,32 +28,75 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
+    /*
+    * tlsInstance is a hack to ensure that only the thread initializing the GC
+    * can use GC functions before the initialization is completed. When it is completed,
+    * the GC object created is atomically stored in the __gshared instance.
+    *
+    * This is necessary because calling thead_init is part of the gc_init routine.
+    * thread_init, in turn, calls GC.enable, GC.disable and GC.malloc (through the new keyword)
+    */
+    static GC tlsInstance;
```

Comment:
That's creates yet another implicit coupling in a quite subtle init order, why not just passing the GC to thread_init directly?
It's all just about a single `new Thread` allocation in `thread_attachThis`.
https://github.com/dlang/druntime/blob/760fd82820dc686436fccb5617a4b8f9aa406ee7/src/core/thread.d#L2138-L2143

Reply:
Better yet, make `thread_init` `@nogc`, see https://github.com/dlang/druntime/pull/2079.